### PR TITLE
Streaming PQ

### PR DESCRIPTION
**Pull request by eddyxu**
Train IVF_PQ on sample dataset, and use streaming to compute PQ code without loading full dataset in memory.

### CODE REVIEW

File: rust/src/arrow/linalg.rs [hunk 1]
```diff
@@ -130,6 +131,15 @@ impl MatrixView {
         }
     }
 
+    pub fn row(&self, i: usize) -> Option<Float32Array> {
+        if i >= self.num_rows() {
+            None
```
Comment (changhiskhan) [inline rust/src/arrow/linalg.rs:136]: Should this be an error?

File: rust/src/dataset.rs [hunk 2]
```diff
@@ -381,16 +379,27 @@ impl Dataset {
                     }
                 }
 
-                let builder = IvfPqIndexBuilder::try_new(
+                let ivf_params = IvfBuildParams {
+                    num_partitions: vec_params.num_partitions as usize,
+                    metric_type: vec_params.metric_type,
+                    max_iters: 50,
```
Comment (changhiskhan) [inline rust/src/dataset.rs:385]: are these values from FAISS?

File: rust/src/dataset.rs [hunk 3]
```diff
@@ -381,16 +379,27 @@ impl Dataset {
...
+                };
+                let pq_params = PQBuildParams {
+                    num_sub_vectors: vec_params.num_sub_vectors as usize,
+                    num_bits: 8,
```
Comment (changhiskhan) [inline rust/src/dataset.rs:389]: do we have to change much to support configurable num_bits? is it just exposing a new API parameter? or is the underlying index creation hard coded to 8 bits?

File: rust/src/index/vector/pq.rs [hunk 4]
```diff
@@ -392,7 +392,18 @@ impl ProductQuantizer {
...
+        Ok(())
+    }
+
+    /// Train a [ProductQuantizer] using an array of vectors.
+    pub async fn fit_transform(
+        &mut self,
+        mat: &MatrixView,
+        metric_type: MetricType,
+    ) -> Result<FixedSizeListArray> {
+        self.train(mat, metric_type, 50).await?;
```
Comment (changhiskhan) [inline rust/src/index/vector/pq.rs:405]: Should we move hard coded constants into one place so we know what needs to change in the future?

File: rust/src/index/vector/ivf.rs [hunk 5]
```diff
@@ -102,11 +99,32 @@ impl<'a> IvfPQIndex<'a> {
         };
         let index_metadata = IvfPQIndexMetadata::try_from(&proto)?;
 
+        let reader_ref = reader.as_ref();
+        let transforms = stream::iter(index_metadata.transforms)
+            .map(|tf| async move {
+                Ok::<Arc<dyn Transformer>, Error>(Arc::new(
+                    OptimizedProductQuantizer::load(
```
Comment (changhiskhan) [inline rust/src/index/vector/ivf.rs:106]: i thought this doesn't use OPQ yet?

**Review by changhiskhan — COMMENTED** [review]
Given how much has changed, could you re-run the sift benchmarks just to make sure the latency / recall profile looks roughly the same still?

File: rust/src/dataset.rs (hunk 3 above)
Comment (eddyxu) [reply rust/src/dataset.rs:389]: `num_nbits` is not configurable via user-facing API yet.  8 bits is just 1 byte. other than `8 bits`, the other configuration needs some work .

File: rust/src/index/vector/pq.rs (hunk 4 above)
Comment (eddyxu) [reply rust/src/index/vector/pq.rs:405]: oh, yea, lemme just make this a parameter passed from user-facing api

**Comment by eddyxu** [comment]
Running on `/benchmark/sift/run_all.sh`
```
Running with ivf512
nprobes: 1, refine=None, recall@10=0.546, mean(s)=0.00124342679977417
nprobes: 1, refine=1, recall@10=0.573, mean(s)=0.001298525333404541
nprobes: 1, refine=5, recall@10=0.780, mean(s)=0.0016234064102172853
nprobes: 1, refine=10, recall@10=0.810, mean(s)=0.0021506214141845705
nprobes: 10, refine=None, recall@10=0.643, mean(s)=0.001989481449127197
nprobes: 10, refine=1, recall@10=0.643, mean(s)=0.002009923458099365
nprobes: 10, refine=5, recall@10=0.968, mean(s)=0.002357909679412842
nprobes: 10, refine=10, recall@10=0.975, mean(s)=0.0028011083602905275
nprobes: 50, refine=None, recall@10=0.627, mean(s)=0.005469377040863037
nprobes: 50, refine=1, recall@10=0.647, mean(s)=0.00498615026473999
nprobes: 50, refine=5, recall@10=0.972, mean(s)=0.005275309085845947
nprobes: 50, refine=10, recall@10=0.995, mean(s)=0.006533384323120117
nprobes: 100, refine=None, recall@10=0.650, mean(s)=0.008666768074035644
nprobes: 100, refine=1, recall@10=0.683, mean(s)=0.009069747924804687
nprobes: 100, refine=5, recall@10=0.973, mean(s)=0.009081656932830811
nprobes: 100, refine=10, recall@10=0.992, mean(s)=0.009473004341125489
```

**Review by changhiskhan — APPROVED** [review]

**Pull request merged**